modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -1700,7 +1706,9 @@ private <T> CompletableFuture<T> appendTxCommand(UUID txId, RequestType cmdType,
                 txOps = new TxCleanupReadyFutureList();
             }
 
-            if (isFinalState(txOps.state)) {
+            TxStateMeta txStateMeta = txManager.stateMeta(txId);
+
+            if (txStateMeta == null || isFinalState(txStateMeta.txState())) {
                 cleanupReadyFut.completeExceptionally(new Exception());
```

Comment:
Seems it's not the line you changed, but I doubt that we should just leave it like that :) Maybe correct exception type and add a message? Or at least create a ticket for that, it's up to you.

I see that another exception was in this code before `new TransactionException(TX_FAILED_READ_WRITE_OPERATION_ERR, "Transaction is already finished."));`